src/pipdeptree/_models/package.py

```diff
@@ -3,18 +3,23 @@
 import re
 from abc import ABC, abstractmethod
 from importlib import import_module
-from importlib.metadata import PackageNotFoundError, metadata, version
+from importlib.metadata import Distribution, PackageNotFoundError, metadata, version
 from inspect import ismodule
 from typing import TYPE_CHECKING
 
-from pip._vendor.pkg_resources import Requirement  # noqa: PLC2701
+from packaging.requirements import Requirement
+from packaging.specifiers import SpecifierSet
 
 if TYPE_CHECKING:
+    from importlib.metadata import Distribution
+
     from pip._internal.metadata import BaseDistribution
-    from pip._vendor.pkg_resources import DistInfoDistribution
 
 
 def pep503_normalize(name: str) -> str:
+    if not name:
+        return ""
+
```

Comment:
I meet the situation that Distribution.name is `None`, which may cause following errors.

Traceback (most recent call last):
  File "/usr/bin/pipdeptree", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/usr/lib/python3.11/site-packages/pipdeptree/__main__.py", line 24, in main
    tree = PackageDAG.from_pkgs(pkgs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/pipdeptree/_models/dag.py", line 40, in from_pkgs
    dist_pkgs = [DistPackage(p) for p in pkgs]
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/pipdeptree/_models/dag.py", line 40, in <listcomp>
    dist_pkgs = [DistPackage(p) for p in pkgs]
                 ^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/pipdeptree/_models/package.py", line 120, in __init__
    super().__init__(obj)
  File "/usr/lib/python3.11/site-packages/pipdeptree/_models/package.py", line 34, in __init__
    self.key = pep503_normalize(obj.name)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/pipdeptree/_models/package.py", line 23, in pep503_normalize
    return re.sub("[-_.]+", "-", name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/re/__init__.py", line 185, in sub
    return _compile(pattern, flags).sub(repl, string, count)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: expected string or bytes-like object, got 'NoneType'

Reply:
Emmm, maybe checking it ahead is a better way.
`dist_pkgs = [DistPackage(p) for p in pkgs if p.name]`

Reply:
Actually, it looks like in older versions of `importlib.metadata` (I'm just scrolling for the different git tags [here](https://github.com/python/cpython/tree/3.8/Lib/importlib)), there is no `Distribution.name()`. However, we can use the `metadata` field since it looks like it's used in Python 3.8.

I'll make the appropriate comments to the code now.

Reply:
... but now that I see you are running Python v3.11 and the `Distribution.name` property should exist (see [here](https://github.com/python/cpython/blob/3.11/Lib/importlib/metadata/__init__.py#L619C1-L622C37)), I have no idea why you are getting `None`. That would mean `self.metadata['Name']` is returning `None` (I also observed this returning `None` when you pass it a bad key like `self.metadata['iAmABadKey']`).

In any case we shouldn't expect a `None` type when we expect name to be type `str`.

Are you able to provide a way to reproduce this?

Reply:
> for dist in dists:
>     if not dist.name:
>         print(dist._path)
Aha, I catch the bad guy with the following code.
It turns out to be an empty dist-info directory, but I have no idea why it occurs.
`/usr/lib64/python3.11/site-packages/python_qpid_proton-0.39.0.dist-info`
I think the empty dist-info directory will cause Distribution.name Distribution.version etc. all return None.
Is this a bug that can be optimized? If so, I think we need to fix that, otherwise we should check it in our code.

Reply:
I downloaded this package and I am not running into any problems getting the name and version metadata:
```console
Python 3.8.18 (default, Dec 19 2023, 17:28:42) 
[GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from importlib.metadata import distribution
>>> d = distribution('python-qpid-proton')
>>> d.metadata['Name']
'python-qpid-proton'
>>> d.metadata['Version']
'0.39.0'
>>> 
```

Reply:
Yes, I tried too, the package is normal. It may be problem of my env? Let's not consider this special case for now, I will remove this.